metadata/moa/scripts/nbconverted/3.align-moa.py

```diff
@@ -0,0 +1,531 @@
+
+    select_cols = [mapping_col, broad_id_col, deprecated_broad_id_col]
+    map_subset_df = map_df.loc[:, select_cols]
+
+    broad_long_df = split_pipe_broad_id(map_subset_df, broad_id_col)
+    all_broad_map.append(broad_long_df)
+
+    if date != "20170327":
+        deprecated_broad_long_df = split_pipe_broad_id(
+            map_subset_df, deprecated_broad_id_col
+        )
+        all_broad_map.append(deprecated_broad_long_df)
+
+# Combine full broad_id map
+all_broad_map = (
+    pd.concat(all_broad_map)
+    # This step drops InChIKeys and Broad_ID combos that exist in multiple dates
+    .drop_duplicates(subset=[mapping_col, "broad_id_unique"]).reset_index(drop=True)
+)
+
+print(all_broad_map.shape)
+all_broad_map.head()
+
+
+# In[9]:
+
+
+# Are there Duplicate InChIKey14s?
+inchi_counts = all_broad_map.loc[:, mapping_col].squeeze().value_counts()
+inchi_counts.hist()
+
+
+# In[10]:
+
+
+# Show an example of the most egregious multi-mapping InChiKey14
+all_broad_map.loc[all_broad_map.loc[:, mapping_col] == inchi_counts.index[0], :]
+
+
+# ## Step 2: Perform the Merge with the Platemap Data
+# 
+# This step identifies the InChIKey14 map for the Cell Painting Repurposing Hub Data Subset.
+
+# In[11]:
+
+
+merged_platemap_df = (
+    all_platemap_df.merge(
+        all_broad_map, left_on="broad_id", right_on="broad_id_unique", how="left"
+    )
+    # We don't care about plate info here
+    .drop_duplicates(subset=["broad_id", mapping_col]).loc[
+        :, ["broad_sample", "broad_id", mapping_col]
+    ]
+    # Drop DMSO
+    .dropna(subset=["broad_sample"])
+)
+
+print(merged_platemap_df.shape)
+merged_platemap_df.head()
+
+
+# In[12]:
+
+
+# Are there Duplicate Broad IDs
+broad_id_counts = merged_platemap_df.broad_id.value_counts()
+broad_id_counts.hist()
+
+
+# In[13]:
+
+
+# There are indeed duplicated broad IDs!
+# This means that some of these broad IDs will recieve alternative MOA/Targets
+duplicate_broad_ids = broad_id_counts[broad_id_counts > 1].index.tolist()
+merged_platemap_df.query("broad_id in @duplicate_broad_ids")
+
+
+# ## Step 3: Align Repurposing Platemap with Most Recent MOA/Target Info
+
+# In[14]:
+
+
+# Load Data
+moa_file = "repurposing_info.tsv"
+moa_df = pd.read_csv(moa_file, sep="\t")
+
+print(moa_df.shape)
+moa_df.head()
+
+
+# In[15]:
+
+
+# Perform the alignment
+platemap_moa_df = merged_platemap_df.merge(
+    moa_df.loc[:, [mapping_col, "pert_iname", "moa", "target"]].drop_duplicates(),
+    on=mapping_col,
+    how="left",
+).loc[:, ["broad_sample", "broad_id", mapping_col, "moa", "target"]]
+
+print(platemap_moa_df.shape)
+platemap_moa_df.head()
+
+
+# In[16]:
+
+
+# Show an example of a broad_id that must be resolved
+platemap_moa_df.query("broad_id == 'BRD-K56844688'")
+
+
+# ## Step 5 - Quantify what's missing
+
+# In[17]:
+
+
+# Some samples are missing an InChIKey14
+platemap_moa_with_inchi_df = platemap_moa_df.dropna(subset=[mapping_col])
```

Comment:
Are these due to some of the `broad_id`s of the profiles no mapping to any `broad_id` in any of the versions?

Reply:
NVM, the last table answered my question.

Reply:
> Are these due to some of the broad_ids of the profiles no mapping to any broad_id in any of the versions?

This was my interpretation. Is this what you deduced from the last table? or something additional?

Reply:
Yes that was my interpretation as well